lib/src/license_detection/tokenizer.dart

```diff
@@ -0,0 +1,112 @@
+import 'dart:collection';
+import 'dart:convert';
+
+class Token {
+  String token;
+  int? position;
+  final int line;
+
+  Token(this.token, this.position, this.line);
+}
+
+class Tokenizer {
+  static const lineSplitter = LineSplitter();
+  static List<Token> tokenize(String text) {
+    var tokens = <Token>[];
+
+    var lines = lineSplitter.convert(text);
+    var position = 0;
+
+    for (var i = 1; i <= lines.length; i++) {
+      var line = lines[i - 1];
+
+      var buffer = StringBuffer();
+      var prevSignificant = false;
+
+      for (var rune in line.runes) {
+        if (!_isDigit(rune) && !prevSignificant) {
+          continue;
+        }
+
+        if (_isSpace(rune)) {
+          if (buffer.isNotEmpty) {
+            tokens.add(Token(buffer.toString(), position++, i));
+            buffer.clear();
+          }
+          prevSignificant = false;
+          continue;
+        }
+
+        buffer.write(String.fromCharCode(rune));
+        prevSignificant = true;
+      }
+      if (buffer.isNotEmpty) {
+        tokens.add(Token(buffer.toString(), position++, i));
+      }
+
+      tokens.add(Token('\n', null, i));
+    }
+
+    //Guideline 7.1.1: Ignore list item for matching purposes.
+    tokens = removeListItems(tokens);
+
+    return tokens;
+  }
+
+  static List<Token> removeListItems(List<Token> tokens) {
+    var newLine = true;
+    var position = 0;
+    var output = <Token>[];
+
+    for (var i = 0; i < tokens.length; i++) {
+      if (newLine && isListItem(tokens[i].token)) {
```

Comment:
This needs documentation and justification... why did we decide to do this? Do others do it, and is it justified in SPDX guidelines? Links...

Reply:
This is what I had to discuss with you. The guideline(6.1.1) says to ignore code comment indicators, this approach will definitely get rid of them but it also remove other punctuation if they are standalone(not attached to words or numbers) or they appear at the start of a word.
For example
`// comment` ---> `comment` (// ignored)
`$dollar` ----> `dollar`
`dolla$r` ----> `dolla$r`

Also further in `v2 classifier`
They ignore everything except `-` `.` `:` in a word.
They say they came to this conclusion on [analysing](https://github.com/google/licenseclassifier/blob/bb04aff29e72e636ba260ec61150c6e15f111d7e/v2/tokenizer.go#L39) the corpus. But I wanted to clarify with you before proceeding.

Reply:
This is one of those examples where I think SPDX guidelines are not practical..

Maybe, we try with the same approach as license classifier v2.
Make a reference to their code and leave a todo to experiment with this..

We can go back and tweak it later... In any case: should this not live in the normalization part?

Reply:
We can't deal with all the details while normalizing, but with tokens, it'll be easier to deal with as we are processing a single word. Hence it is here.